File: Sources/DebridStreamer/Services/Sync/IMDbCSVSyncService.swift
import Foundation

struct IMDbCSVEntry: Sendable, Equatable {
    var imdbID: String?
    var title: String
    var year: Int?
    var listType: UserLibraryEntry.ListType
}

struct IMDbImportResult: Sendable, Equatable {
    var added: Int
    var skippedDuplicates: Int
}

actor IMDbCSVSyncService {
    /// Builds a stable, diacritic/case-insensitive slug from a title for use as a
    /// synthesized media id when no IMDb const is present. Collapses runs of
    /// non-alphanumeric characters to single hyphens and strips diacritics, so the
    /// raw title text never leaks into the id and casing/accents don't fork ids.
    nonisolated func normalizedSlug(_ s: String) -> String {
        s.folding(options: [.diacriticInsensitive, .caseInsensitive], locale: nil)
            .components(separatedBy: CharacterSet.alphanumerics.inverted)
            .filter { !$0.isEmpty }
            .joined(separator: "-")
    }

    nonisolated func parseCSV(_ contents: String, listType: UserLibraryEntry.ListType) -> [IMDbCSVEntry] {
        let rows = CSVParser.parse(contents: contents)
        guard let header = rows.first else { return [] }
        let headerIndex = Dictionary(uniqueKeysWithValues: header.enumerated().map { ($1.lowercased(), $0) })

        let idxConst = headerIndex["const"]
        let idxTitle = headerIndex["title"]
        let idxYear = headerIndex["year"]

        return rows.dropFirst().compactMap { row in
            guard let idxTitle, idxTitle < row.count else { return nil }
            let title = row[idxTitle].trimmingCharacters(in: .whitespacesAndNewlines)
            guard !title.isEmpty else { return nil }

            var imdbID: String?
            if let idxConst, idxConst < row.count {
                let value = row[idxConst].trimmingCharacters(in: .whitespacesAndNewlines)
                imdbID = value.isEmpty ? nil : value
            }

            var year: Int?
            if let idxYear, idxYear < row.count {
                year = Int(row[idxYear].trimmingCharacters(in: .whitespacesAndNewlines))
            }

            return IMDbCSVEntry(imdbID: imdbID, title: title, year: year, listType: listType)
        }
    }

    func importCSV(
        _ contents: String,
        listType: UserLibraryEntry.ListType,
        folderId: String,
        database: DatabaseManager
    ) async throws -> IMDbImportResult {
        let entries = parseCSV(contents, listType: listType)
        var added = 0
        var skipped = 0

        // Pre-compute the synthesized media id for every row once, so the bulk reads
        // below and the per-row inserts agree on a single id per entry.
        let mediaIDs = entries.map { entry in
            entry.imdbID ?? "imdb-\(normalizedSlug(entry.title))-\(entry.year ?? 0)"
        }

        // Bulk read: which of these media already exist in the cache. Replaces the
        // per-row fetchMedia(id:) round-trips with one batched query.
        let uniqueIDs = Array(Set(mediaIDs))
        let existingMedia = try await database.fetchMedia(ids: uniqueIDs)

        // Track media inserted earlier in this same import so we don't re-save them,
        // mirroring the original per-row "fetchMedia == nil" guard within the batch.
        var savedMediaIDs = Set(existingMedia.keys)

        for (index, entry) in entries.enumerated() {
            let mediaID = mediaIDs[index]
            let libraryID = "\(mediaID)-\(folderId)"

            // Preserve the original dedup semantics exactly: check against the raw
            // folderId per row (same as the previous isInLibrary(mediaId:folderId:)).
            let exists = try await database.isInLibrary(mediaId: mediaID, folderId: folderId)
            if exists {
                skipped += 1
                continue
            }

            if !savedMediaIDs.contains(mediaID) {
                let media = MediaItem(
                    id: mediaID,
                    type: .movie,
                    title: entry.title,
                    year: entry.year,
                    lastFetched: Date()
                )
                try await database.saveMedia(media)
                savedMediaIDs.insert(mediaID)
            }

            let libraryEntry = UserLibraryEntry(
                id: libraryID,
                mediaId: mediaID,
                folderId: folderId,
                listType: listType,
                addedAt: Date()
            )
            try await database.addToLibrary(libraryEntry)
            added += 1
        }

        return IMDbImportResult(added: added, skippedDuplicates: skipped)
    }

    func importCSV(
        _ contents: String,
        listType: UserLibraryEntry.ListType,
        database: DatabaseManager
    ) async throws -> IMDbImportResult {
        let folderId = try await database.fetchSystemLibraryFolderID(listType: listType)
        return try await importCSV(
            contents,
            listType: listType,
            folderId: folderId,
            database: database
        )
    }

    nonisolated func exportCSV(mediaItems: [MediaItem]) -> String {
        var rows: [[String]] = [["Const", "Title", "Year"]]
        rows.append(contentsOf: mediaItems.map { item in
            [item.id, item.title, item.year.map(String.init) ?? ""]
        })
        return CSVParser.serialize(rows: rows)
    }

    func exportCSV(
        database: DatabaseManager,
        folderId: String,
        includeDescendants: Bool = true
    ) async throws -> String {
        let mediaItems = try await database.fetchLibraryMedia(
            folderId: folderId,
            includeDescendants: includeDescendants
        )
        return exportCSV(mediaItems: mediaItems)
    }

    func exportCSVAllFolders(
        database: DatabaseManager,
        rootFolderId: String? = nil
    ) async throws -> String {
        let mediaItems = try await database.fetchLibraryMediaInFolderTree(rootFolderId: rootFolderId)
        return exportCSV(mediaItems: mediaItems)
    }
}

private enum CSVParser {
    static func parse(contents: String) -> [[String]] {
        var rows: [[String]] = []
        var row: [String] = []
        var cell = ""
        var inQuotes = false

        let chars = Array(contents)
        var i = 0
        while i < chars.count {
            let ch = chars[i]
            if ch == "\"" {
                if inQuotes, i + 1 < chars.count, chars[i + 1] == "\"" {
                    cell.append("\"")
                    i += 1
                } else {
                    inQuotes.toggle()
                }
            } else if ch == "," && !inQuotes {
                row.append(cell)
                cell = ""
            } else if (ch == "\n" || ch == "\r") && !inQuotes {
                if ch == "\r", i + 1 < chars.count, chars[i + 1] == "\n" {
                    i += 1
                }
                row.append(cell)
                if !row.allSatisfy({ $0.isEmpty }) {
                    rows.append(row)
                }
                row = []
                cell = ""
            } else {
                cell.append(ch)
            }
            i += 1
        }

        if !cell.isEmpty || !row.isEmpty {
            row.append(cell)
            if !row.allSatisfy({ $0.isEmpty }) {
                rows.append(row)
            }
        }

        return rows
    }

    static func serialize(rows: [[String]]) -> String {
        rows.map { row in
            row.map(escape).joined(separator: ",")
        }.joined(separator: "\n")
    }

    private static func escape(_ value: String) -> String {
        if value.contains(",") || value.contains("\"") || value.contains("\n") {
            return "\"\(value.replacingOccurrences(of: "\"", with: "\"\""))\""
        }
        return value
    }
}

File: web/src/services/sync/IMDbCSVSyncService.ts
// Port of Sources/DebridStreamer/Services/Sync/IMDbCSVSyncService.swift.
//
// Ports the PURE HTTP-free CSV logic: the RFC-4180-ish CSV parser/serializer,
// `normalizedSlug`, `parseCSV`, `exportCSV(mediaItems:)`, and the synthesized
// fallback media-id rule used during import. The two `importCSV(...)` and
// `exportCSV(database:)`/`exportCSVAllFolders(...)` methods are DB-backed
// (they take a `DatabaseManager` and read/write rows); that DB layer is not yet
// ported to web, so those orchestration methods are intentionally OMITTED here.
// The id-synthesis the importer relies on is exposed as `fallbackMediaID` so
// the dedup-key logic is unit-tested independently of the DB.

import type { MediaItem } from "../../models/media";
import type { IMDbCSVEntry, ListType } from "./models";

export class IMDbCSVSyncService {
  /**
   * Builds a stable, diacritic/case-insensitive slug from a title for use as a
   * synthesized media id when no IMDb const is present. Collapses runs of
   * non-alphanumeric characters to single hyphens and strips diacritics, so the
   * raw title text never leaks into the id and casing/accents don't fork ids.
   * Mirrors Swift `normalizedSlug`.
   */
  normalizedSlug(s: string): string {
    // NFD + strip combining marks == Swift's `.diacriticInsensitive` folding.
    const deaccented = s.normalize("NFD").replace(/\p{M}/gu, "");
    return deaccented
      .toLowerCase()
      // Split on any run of non-alphanumerics (Swift:
      // CharacterSet.alphanumerics.inverted). \p{L}\p{N} keeps Unicode letters/
      // digits, matching Swift's Unicode-aware alphanumerics.
      .split(/[^\p{L}\p{N}]+/u)
      .filter((part) => part.length > 0)
      .join("-");
  }

  /**
   * Synthesized media id for a parsed entry: the IMDb const when present, else
   * `imdb-{slug}-{year}` (year defaults to 0). Mirrors the per-row id the Swift
   * `importCSV` computes for dedup/insert. Pure so the dedup key is testable
   * without a DB.
   */
  fallbackMediaID(entry: IMDbCSVEntry): string {
    if (entry.imdbID != null) return entry.imdbID;
    return `imdb-${this.normalizedSlug(entry.title)}-${entry.year ?? 0}`;
  }

  /**
   * Parses an IMDb-style CSV into entries. Reads the `Const`/`Title`/`Year`
   * columns case-insensitively from the header, skips rows with an empty title,
   * and trims cells. Mirrors Swift `parseCSV`.
   */
  parseCSV(contents: string, listType: ListType): IMDbCSVEntry[] {
    const rows = CSVParser.parse(contents);
    const header = rows[0];
    if (header == null) return [];

    const headerIndex = new Map<string, number>();
    header.forEach((name, index) => {
      headerIndex.set(name.toLowerCase(), index);
    });

    const idxConst = headerIndex.get("const");
    const idxTitle = headerIndex.get("title");
    const idxYear = headerIndex.get("year");

    const entries: IMDbCSVEntry[] = [];
    for (const row of rows.slice(1)) {
      if (idxTitle == null || idxTitle >= row.length) continue;
      const title = row[idxTitle].trim();
      if (title.length === 0) continue;

      let imdbID: string | null = null;
      if (idxConst != null && idxConst < row.length) {
        const value = row[idxConst].trim();
        imdbID = value.length === 0 ? null : value;
      }

      let year: number | null = null;
      if (idxYear != null && idxYear < row.length) {
        const parsed = Number.parseInt(row[idxYear].trim(), 10);
        // Swift `Int(_:)` rejects trailing junk; mirror that with a strict check.
        year = /^[+-]?\d+$/.test(row[idxYear].trim()) && !Number.isNaN(parsed)
          ? parsed
          : null;
      }

      entries.push({ imdbID, title, year, listType });
    }
    return entries;
  }

  /**
   * Serializes media items into an IMDb-compatible CSV with a
   * `Const,Title,Year` header. Mirrors Swift `exportCSV(mediaItems:)`.
   */
  exportCSV(mediaItems: MediaItem[]): string {
    const rows: string[][] = [["Const", "Title", "Year"]];
    for (const item of mediaItems) {
      rows.push([item.id, item.title, item.year != null ? String(item.year) : ""]);
    }
    return CSVParser.serialize(rows);
  }
}

// MARK: - CSV parser/serializer (mirrors the private Swift `CSVParser`)

const CSVParser = {
  /** RFC-4180-ish parse: handles quoted fields, escaped `""`, and CR/LF/CRLF
   * row terminators. Drops all-empty rows. Mirrors Swift `CSVParser.parse`. */
  parse(contents: string): string[][] {
    const rows: string[][] = [];
    let row: string[] = [];
    let cell = "";
    let inQuotes = false;

    const chars = Array.from(contents);
    let i = 0;
    while (i < chars.length) {
      const ch = chars[i];
      if (ch === '"') {
        if (inQuotes && i + 1 < chars.length && chars[i + 1] === '"') {
          cell += '"';
          i += 1;
        } else {
          inQuotes = !inQuotes;
        }
      } else if (ch === "," && !inQuotes) {
        row.push(cell);
        cell = "";
      } else if ((ch === "\n" || ch === "\r") && !inQuotes) {
        if (ch === "\r" && i + 1 < chars.length && chars[i + 1] === "\n") {
          i += 1;
        }
        row.push(cell);
        if (!row.every((c) => c.length === 0)) {
          rows.push(row);
        }
        row = [];
        cell = "";
      } else {
        cell += ch;
      }
      i += 1;
    }

    if (cell.length > 0 || row.length > 0) {
      row.push(cell);
      if (!row.every((c) => c.length === 0)) {
        rows.push(row);
      }
    }

    return rows;
  },

  /** Serializes rows, quoting any cell containing `,`, `"`, or newline and
   * doubling embedded quotes. Mirrors Swift `CSVParser.serialize`. */
  serialize(rows: string[][]): string {
    return rows.map((row) => row.map(escapeCSV).join(",")).join("\n");
  },
};

function escapeCSV(value: string): string {
  if (value.includes(",") || value.includes('"') || value.includes("\n")) {
    return `"${value.replace(/"/g, '""')}"`;
  }
  return value;
}
